pkg/runtime/sequencer/cache.go

```diff
@@ -15,15 +15,203 @@ package sequencer
 
 import (
 	"context"
+	"errors"
 	"mosn.io/layotto/components/sequencer"
+	"mosn.io/pkg/log"
+	"mosn.io/pkg/utils"
+	"sync"
+	"time"
 )
 
+const defaultSize = 10000
+const defaultLimit = 1000
+const defaultRetry = 5
+const waitTime = time.Second * 2
+
+// DoubleBuffer is double segment id buffer.
+// There are two buffers in DoubleBuffer: inUseBuffer is in use, BackUpBuffer is a backup buffer.
+// Their default capacity is 1000. When the inUseBuffer usage exceeds 30%, the BackUpBuffer will be initialized.
+// When inUseBuffer is used up, swap them.
+type DoubleBuffer struct {
+	Key              string
+	size             int
+	inUseBuffer      *Buffer
+	backUpBufferChan chan *Buffer
+	lock             sync.Mutex
+	Store            sequencer.Store
+}
+
+type Buffer struct {
+	from int64
+	to   int64
+}
+
+func NewDoubleBuffer(key string, store sequencer.Store) *DoubleBuffer {
+
+	d := &DoubleBuffer{
+		Key:              key,
+		size:             defaultSize,
+		Store:            store,
+		backUpBufferChan: make(chan *Buffer, 1),
+	}
+
+	return d
+}
+
+//init double buffer
+func (d *DoubleBuffer) init() error {
+
+	buffer, err := d.getNewBuffer()
+	if err != nil {
+		return err
+	}
+
+	d.inUseBuffer = buffer
+
+	return nil
+}
+
+//getId next id
+func (d *DoubleBuffer) getId() (int64, error) {
+
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.inUseBuffer == nil {
+		return 0, errors.New("[DoubleBuffer] Get error: inUseBuffer nil ")
+	}
+	//check swap
+	if d.inUseBuffer.from > d.inUseBuffer.to {
+		err := d.swap()
+		if err != nil {
+			return 0, err
+		}
+	}
+	next := d.inUseBuffer.from
+	d.inUseBuffer.from++
+
+	//when inUseBuffer id more than limit used, initialize BackUpBuffer.
+	//equal make sure only one thread enter
+	if d.inUseBuffer.to-d.inUseBuffer.from == defaultLimit {
+		utils.GoWithRecover(func() {
+
+			//retry get buffer
+			for i := 0; i < defaultRetry; i++ {
```

Comment:
shoudl loop forever until success

otherwise, once defaultRetry exceeded, getId() alwasy return err

Reply:
when  error occurs,  need to wait some time before trying again ?